course-template-bundle-spec.md

```diff
@@ -179,6 +181,46 @@ attribute        | required | type        | notes
 id               | ✓        | string      | A unique identifier for this step
 activity_options | ✓        | array       | `activity_options` is an array of dictionaries with the format:
 -- type          | ✓        | enum        | One of `lab`, `quiz`, `resource`
+-- subtype       |          | string      | A subordinate activity type; allowed `subtype` strings are contingent on the above `type`
 -- id            | ✓        | string      | Reference to the unique identifier for the activity - `library/slug`.
 prompt           |          | dictionary  | Key is `locales` and each locale is a dictionary mapping locale codes to a prompt describing the step
 optional         |          | boolean     | `true` if the step is *not* required for completion
+
+### Retake Policies
+
+A retake policy defines any required cooldown periods, retake limits, and retake windows applicable to a given `CourseTemplate` activity.
+
+The overall format should look like:
+
+```yml
+retake_policies:
+  - id: default-lab-policy
+    activity_type: Lab
+    retake_cooldown:
+      - 1
+      - 2
+      - 5
+
+  - id: practice-quiz-policy
+    activity_type: Quiz
+    activity_subtype: practice
+    retake_limit: 3
+    retake_window: 1
+
+  - id: graded-quiz-policy
+    activity_type: Quiz
+    activity_subtype: graded
+    retake_cooldown:
+      - 1
+    retake_limit: 3
+    retake_window: 1
+```
+
+attribute               | required | type        | notes
+----------------------- | -------- | ----------- | -----------------------------------------
+id                      | ✓        | string      | A unique identifier for this retake policy
```

Comment:
If we're matching on `activity_type` and `activity_subtype`, we can only have one retake policy per combo, which makes this `id` field redundant... or at least simply decorative.

"Decorative" properties can absolutely be valuable in and of themselves if they act as an affordances for users; but the ContentBundle spec is targeted for machines, not humans. So I'd ask for a light justification about future use cases to keep the id around.

Alternatively, we could explicitly reference `retake_policy_id` on activities, and eschew the `subtype`.

Reply:
The presence of the additional ID is really to satisfy this check here: https://github.com/CloudVLab/git-whisperer/blob/master/whisperer/utils/yaml_merger.py#L134, less so because retaking requires it.

Reply:
Ah, that's a good call in general; however...

1. Confusingly, it may be appropriate for "GitWhisperer format" to have an ID specifically for locale merging, but we don't need to include it as part of the Content Bundle spec. It's likely a GitWhisperer specific implementation detail. [Documentation for the GitWhisperer format is sorely needed as well.]

2. It doesn't look like any of these properties need to be localized, so we they don't need to be included in the locale specific yaml files (or merged).

Reply:
yeah that's what i thought too, but carine was saying she saw the error, and b/160686692 makes me wonder if there's a bug in that as well...

Reply:
If that the behavior is intended to be (and will be fixed to be) such that non-localized objects in an array don't require an `id`, then SG, I will remove this `id` in the follow-up PR as well.